### PR TITLE
[SHELL32] Commonize CCopyToMenu and CMoveToMenu

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Reduce code. And correctly do the filename check.

JIRA issue: [CORE-18909](https://jira.reactos.org/browse/CORE-18909)

## TODO

- [x] Do tests.

### CODE REVIEW

**Review by JoachimHenze — APPROVED** [review]
I do slightly remember that some devs like Timo Kreuzer expressed their dislike in that the control flow is controlled by the title here. But personally I do have no problem with it.

I do appreciate the small efficiency tweaks that Katayama applied here on top of the initially proposed patch, e.g. relying on the default constructor to create an empty string, and using .IsEmpty() method instead of an explicit string comparison.

**Comment by whindsaks** [comment]
CCopyToMenu::DoGetFileTitle and CMoveToMenu::DoGetFileTitle are exactly the same, should it not exist as a helper function?

This probably applies to everything here. What is the difference between these two files other than FO_MOVE vs FO_COPY and some minor menu and resource id stuff?

This could exist as a single implementation that either provides both in one COM object or as a c++ template.

**Comment by TAN-Gaming** [comment]
I am not sure this is the correct fix. See comments from @learn-more on my PR: [https://github.com/reactos/reactos/pull/5235](https://github.com/reactos/reactos/pull/5235).

**Comment by katahiromz** [comment]
VirtualBox_ReactOS_03_11_2023_15_42_01
It's still working...

File: dll/win32/shell32/CCopyToMoveToMenu.cpp [hunk 1]
```diff
@@ -1,33 +1,29 @@
...
+ * PURPOSE:     CopyTo and MoveTo implementation
+ * COPYRIGHT:   Copyright 2020-2023 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
  */
 
 #include "precomp.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(shell);
 
-CMoveToMenu::CMoveToMenu() :
+CCopyToMoveToMenu::CCopyToMoveToMenu() :
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMoveToMenu.cpp]: OK.

File: dll/win32/shell32/CCopyToMoveToMenu.cpp [hunk 2]
```diff
@@ -1,33 +1,29 @@
...
     m_idCmdFirst(0),
     m_idCmdLast(0),
-    m_idCmdMoveTo(-1),
+    m_idCmdDoTo(-1),
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMoveToMenu.cpp]: OK.

File: dll/win32/shell32/CCopyToMoveToMenu.cpp [hunk 3]
```diff
@@ -64,30 +60,30 @@ WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
     return CallWindowProcW(this_->m_fnOldWndProc, hwnd, uMsg, wParam, lParam);
 }
 
-static int CALLBACK
-BrowseCallbackProc(HWND hwnd, UINT uMsg, LPARAM lParam, LPARAM lpData)
+INT CALLBACK
+CCopyToMoveToMenu::BrowseCallbackProc(HWND hwnd, UINT uMsg, LPARAM lParam, LPARAM lpData)
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMoveToMenu.cpp]: Made them global functions.

**Comment by HBelusca** [comment]
yes i forgot to approve but it was OK :+1:

**Pull request merged**